include/slicerecon/reconstruction/reconstructor.hpp

```diff
@@ -165,56 +174,86 @@ class reconstructor {
+
+                        if (end_wrt_geom > begin_wrt_geom) {
+                            upload_sino_buffer_(begin_wrt_geom, end_wrt_geom, 0, use_gpu_lock);
+                        } else {
+                            // we have gone around in the geometry
+                            upload_sino_buffer_(begin_wrt_geom, geom_.proj_count - 1, 0, use_gpu_lock);
+                            upload_sino_buffer_(0, end_wrt_geom, geom_.proj_count - begin_wrt_geom - 1, use_gpu_lock);
+                        }
+
+                        if (is_alt) write_index_ = 1 - write_index_; // swap buffers
+                        refresh_data_();
 
-            auto start_idx = start_index_(idx);
-            memcpy(&buffer_[write_index_][start_idx], data,
-                   sizeof(float) * pixels_);
-
-            if (idx % parameters_.group_size == parameters_.group_size - 1 ||
-                idx == geom_.proj_count - 1) {
-                // upload and switch writing idx
-                upload_(
-                    current_group_ * parameters_.group_size,
-                    std::min((current_group_ + 1) * parameters_.group_size - 1,
-                             geom_.proj_count - 1));
-                current_group_ = (current_group_ + 1) % group_count_;
-
-                if (current_group_ == 0) {
-                    refresh_data_();
-                    write_index_ = 1 - write_index_;
+                        update_count_++;
+                    }
+
+                    current_group_ = (current_group_ + 1) % group_count_;
                 }
-            }
 
-            break;
-        }
-        case proj_kind::dark: {
-            memcpy(&all_darks_[idx * pixels_], data, sizeof(float) * pixels_);
-            received_flats_++;
-            break;
-        }
-        case proj_kind::light: {
-            memcpy(&all_flats_[idx * pixels_], data, sizeof(float) * pixels_);
-            received_flats_++;
-            break;
-        }
-        default:
-            break;
+                break;
+            }
+            case proj_kind::dark: {
+                memcpy(&all_darks_[proj_idx * pixels_], data, sizeof(float) * pixels_);
+                received_flats_++;
+                break;
+            }
+            case proj_kind::light: {
+                memcpy(&all_flats_[proj_idx * pixels_], data, sizeof(float) * pixels_);
+                received_flats_++;
+                break;
+            }
+            default:
+                break;
         }
     }
 
     slice_data reconstruct_slice(orientation x) {
+        // the lock is supposed to be always open if reconstruction mode == alternating
+        std::lock_guard<std::mutex> guard(gpu_mutex_);
+
         if (!initialized_) {
             return {{1, 1}, {0.0f}};
         }
 
-        return alg_->reconstruct_slice(x, 1 - write_index_);
+        return alg_->reconstruct_slice(x, write_index_);
```

Comment:
Is this still correct for `alternating` mode, where we want to have one _write index_ and one _reconstruction index_?

Reply:
I have now substituted this in favor for `active_gpu_buffer_index_`.